File: backend/game.py
from board import create_board, clone_board, board_hash, BLACK, WHITE
from rules import is_legal_move, apply_move, get_legal_moves
from score import compute_score

class Game:
    def __init__(self):
        self.board = create_board()
        self.current_player = BLACK
        self.captures = {BLACK: 0, WHITE: 0}
        self.prev_board_hash = None
        self.board_history = set()
        self.last_move = None
        self.game_over = False
        self.consecutive_passes = 0
        self._undo_stack = []

    def _snapshot(self):
        return {
            'board': self.board.copy(),
            'current_player': self.current_player,
            'captures': dict(self.captures),
            'prev_board_hash': self.prev_board_hash,
            'board_history': set(self.board_history),
            'last_move': self.last_move,
            'game_over': self.game_over,
            'consecutive_passes': self.consecutive_passes,
        }

    def _restore(self, snap):
        self.board = snap['board']
        self.current_player = snap['current_player']
        self.captures = snap['captures']
        self.prev_board_hash = snap['prev_board_hash']
        self.board_history = snap['board_history']
        self.last_move = snap['last_move']
        self.game_over = snap['game_over']
        self.consecutive_passes = snap['consecutive_passes']

    def play_move(self, row, col):
        if self.game_over:
            return False
        if not is_legal_move(self.board, row, col, self.current_player,
                              self.board_history or self.prev_board_hash):
            return False

        self._undo_stack.append(self._snapshot())
        self.board_history.add(board_hash(self.board))
        self.prev_board_hash = board_hash(self.board)
        self.board, captured = apply_move(self.board, row, col, self.current_player)
        self.captures[self.current_player] += captured
        self.last_move = (row, col)
        self.consecutive_passes = 0
        self.current_player = 3 - self.current_player
        return True

    def pass_turn(self):
        """Pass this turn. Game ends when both players pass consecutively."""
        if self.game_over:
            return
        self._undo_stack.append(self._snapshot())
        self.consecutive_passes += 1
        self.last_move = None
        if self.consecutive_passes >= 2:
            self.game_over = True
        else:
            self.current_player = 3 - self.current_player

    def undo(self):
        """Undo the last move or pass. Returns True if successful."""
        if not self._undo_stack:
            return False
        self._restore(self._undo_stack.pop())
        return True

    def get_board(self):
        return self.board

    def get_legal_moves(self):
        return get_legal_moves(self.board, self.current_player,
                               self.board_history or self.prev_board_hash)

    def get_score(self):
        return compute_score(self.board)

    def is_game_over(self):
        return self.game_over

File: cpp_engine/src/game.cpp
#include "game.h"
#include "logger.h"
#include <cstdlib>
#include <iostream>
#include <sstream>
#include <algorithm>
#include <random>

Game::Game() {
    board_history.reserve(256);
}

bool Game::play_move(int row, int col) {
    if (game_over) return false;
    if (!is_legal_move(board, row, col, current_player, board_history)) return false;

    board_history.insert(board.hash());
    int captured = apply_move(board, row, col, current_player);
    captures[current_player] += captured;

    last_move_index  = row * BOARD_SIZE + col;
    consecutive_passes = 0;
    ++move_number;
    current_player = (int8_t)(3 - current_player);
    return true;
}

void Game::pass_turn() {
    if (game_over) return;
    board_history.insert(board.hash());
    last_move_index = -1;
    ++consecutive_passes;
    ++move_number;
    if (consecutive_passes >= 2) {
        game_over = true;
    } else {
        current_player = (int8_t)(3 - current_player);
    }
}

std::vector<int> Game::get_legal_moves() const {
    return ::get_legal_moves(board, current_player, board_history);
}

void Game::get_legal_moves_mask(std::array<bool, TOTAL_CELLS>& mask) const {
    ::get_legal_moves_mask(board, current_player, board_history, mask);
}

ScoreResult Game::get_score() const {
    return compute_score(board);
}

// 4-channel encoding matching integrate_cnn.py:
//   ch0 = black stones, ch1 = white stones
//   ch2 = current_player-1 plane (0.0 if BLACK, 1.0 if WHITE)
//   ch3 = legal moves mask
void Game::encode_board(float* out, int /*num_channels*/) const {
    float cp_plane = (float)(current_player - 1);  // 0 if BLACK, 1 if WHITE

    std::array<bool, TOTAL_CELLS> legal_mask;
    get_legal_moves_mask(legal_mask);

    for (int idx = 0; idx < TOTAL_CELLS; ++idx) {
        int8_t v = board.get_idx(idx);
        out[0 * TOTAL_CELLS + idx] = (v == BLACK) ? 1.0f : 0.0f;  // ch0 black
        out[1 * TOTAL_CELLS + idx] = (v == WHITE) ? 1.0f : 0.0f;  // ch1 white
        out[2 * TOTAL_CELLS + idx] = cp_plane;                      // ch2 current player
        out[3 * TOTAL_CELLS + idx] = legal_mask[idx] ? 1.0f : 0.0f;// ch3 legal
    }
}

int8_t Game::random_rollout(int max_moves) const {
    Game tmp = clone();
    thread_local std::mt19937 rng(std::random_device{}());

    int safety = 0;
    while (!tmp.game_over && safety < max_moves) {
        ++safety;
        auto moves = tmp.get_legal_moves();
        if (moves.empty()) {
            tmp.pass_turn();
        } else {
            std::uniform_int_distribution<int> dist(0, (int)moves.size() - 1);
            int m = moves[dist(rng)];
            tmp.play_move(m / BOARD_SIZE, m % BOARD_SIZE);
        }
    }
    auto sc = tmp.get_score();
    return sc.winner;
}

Game Game::clone() const {
    return *this;   // board_history, board, etc. all value-copy
}

std::string Game::to_string() const {
    std::ostringstream oss;
    oss << board.to_string();
    oss << "Player: " << (current_player == BLACK ? "BLACK" : "WHITE")
        << "  Move: " << move_number
        << "  B-cap: " << captures[BLACK] << "  W-cap: " << captures[WHITE] << '\n';
    return oss.str();
}

void Game::print_board() const {
    std::cout << to_string();
}
